extensions/typescript/src/features/referencesCodeLensProvider.ts

```diff
@@ -0,0 +1,118 @@
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import { CodeLensProvider, CodeLens, CancellationToken, TextDocument, Range, Uri, Location, Position } from 'vscode';
+import * as Proto from '../protocol';
+import * as PConst from '../protocol.const';
+
+import { ITypescriptServiceClient } from '../typescriptService';
+
+import * as nls from 'vscode-nls';
+let localize = nls.loadMessageBundle();
+
+
+class ReferencesCodeLens extends CodeLens {
+	public document: Uri;
+	public file: string;
+
+	constructor(document: Uri, file: string, range: Range) {
+		super(range);
+		this.document = document;
+		this.file = file;
+	}
+}
+
+export default class TypeScriptReferencesCodeLensProvider implements CodeLensProvider {
+	private client: ITypescriptServiceClient;
+
+	constructor(client: ITypescriptServiceClient) {
+		this.client = client;
+	}
+
+	provideCodeLenses(document: TextDocument, token: CancellationToken): Promise<CodeLens[]> {
+		const filepath = this.client.asAbsolutePath(document.uri);
+		if (!filepath) {
+			return Promise.resolve([]);
+		}
+		return this.client.execute('navtree', { file: filepath }, token).then(response => {
+			const tree = response.body;
+			const referenceableSpans: Range[] = [];
+			if (tree && tree.childItems) {
+				tree.childItems.forEach(item => this.extractReferenceableSymbols(item, referenceableSpans));
+			}
+			return Promise.resolve(referenceableSpans.map(span => new ReferencesCodeLens(document.uri, filepath, span)));
+		});
+	}
+
+	resolveCodeLens(inputCodeLens: CodeLens, token: CancellationToken): Promise<CodeLens> {
+		const codeLens = inputCodeLens as ReferencesCodeLens;
+		if (!codeLens.document) {
+			return Promise.reject<CodeLens>(codeLens);
+		}
+		const args: Proto.FileLocationRequestArgs = {
+			file: codeLens.file,
+			line: codeLens.range.start.line + 1,
+			offset: codeLens.range.start.character + 1
+		};
+		return this.client.execute('references', args, token).then(response => {
+			if (response && response.body) {
+				const referenceCount = Math.max(0, response.body.refs.length - 1);
+				const locations = response.body.refs.map(reference =>
+					new Location(codeLens.document,
+						new Range(
+							new Position(reference.start.line - 1, reference.start.offset - 1),
+							new Position(reference.end.line - 1, reference.end.offset - 1))));
+
+				codeLens.command = {
+					title: referenceCount + ' ' + (referenceCount === 1 ? localize('oneReferenceLabel', 'reference') : localize('manyReferenceLabel', 'references')),
+					command: 'editor.action.showReferences',
+					arguments: [codeLens.document, locations[0].range.start, locations.slice(1)]
+				};
+				return Promise.resolve(codeLens);
+			}
+			return Promise.reject(codeLens);
+		}).catch(() => {
+			codeLens.command = {
+				title: localize('referenceErrorLabel', 'Could not determine references'),
+				command: ''
+			};
+			return Promise.resolve(codeLens);
```

Comment:
I am afraid you cannot. The idea is that the editor does shake while we resolve things, e.g in the first, hopefully quick, phase we create all the empty space which make the editor 'jump' and then in the second phase we fill in those holes. Two options

* know in the first phase which lenses you cannot fill in (this the very much preferred solution), or
* use the new API of being  able to tell the editor your set of lenses has [changed](https://github.com/Microsoft/vscode/blob/master/src/vs/vscode.d.ts#L1619). Will cause jumping/flicker and still requires you know valid lenses in the first phase